src/Module/OStatus/PubSub.php

```diff
@@ -0,0 +1,156 @@
+
+		$this->database = $database;
+	}
+
+	protected function post(array $request = [])
+	{
+		$xml = Network::postdata();
+
+		$this->logger->info('Feed arrived from ' . $this->server['REMOTE_ADDR'] . ' for ' . $this->args->getCommand() . ' with user-agent: ' . $this->server['HTTP_USER_AGENT']);
+		$this->logger->debug('Data: ' . $xml);
+
+		$nickname   = $this->parameters['nickname'] ?? '';
+		$contact_id = $this->parameters['cid']      ?? 0;
+
+		$importer = $this->database->selectFirst('user', [], ['nickname' => $nickname, 'account_expired' => false, 'account_removed' => false]);
+		if (!$importer) {
+			throw new HTTPException\OKException();
+		}
+
+		$condition = ['id' => $contact_id, 'uid' => $importer['uid'], 'subhub' => true, 'blocked' => false];
+		$contact   = $this->database->selectFirst('contact', [], $condition);
+		if (!$contact) {
+			$author = OStatus::salmonAuthor($xml, $importer);
+			if (!empty($author['contact-id'])) {
+				$condition = ['id' => $author['contact-id'], 'uid' => $importer['uid'], 'subhub' => true, 'blocked' => false];
+				$contact   = $this->database->selectFirst('contact', [], $condition);
+				$this->logger->notice('No record for ' . $nickname . ' with contact id ' . $contact_id . ' - using ' . $author['contact-id'] . ' instead.');
+			}
+
+			if (!$contact) {
+				$this->logger->notice('Contact ' . $author["author-link"] . ' (' . $contact_id . ') for user ' . $nickname . " wasn't found - ignored. XML: " . $xml);
+				throw new HTTPException\OKException();
+			}
+		}
+
+		if (!empty($contact['gsid'])) {
+			GServer::setProtocol($contact['gsid'], Post\DeliveryData::OSTATUS);
+		}
+
+		if (!in_array($contact['rel'], [Contact::SHARING, Contact::FRIEND]) && ($contact['network'] != Protocol::FEED)) {
+			$this->logger->notice('Contact ' . $contact['id'] . ' is not expected to share with us - ignored.');
+			throw new HTTPException\OKException();
+		}
+
+		// We only import feeds from OStatus here
+		if (!in_array($contact['network'], [Protocol::ACTIVITYPUB, Protocol::OSTATUS])) {
+			$this->logger->warning('Unexpected network', ['contact' => $contact]);
+			throw new HTTPException\OKException();
+		}
+
+		$this->logger->info('Import item for ' . $nickname . ' from ' . $contact['nick'] . ' (' . $contact['id'] . ')');
+		$feedhub = '';
+		OStatus::import($xml, $importer, $contact, $feedhub);
+
+		throw new HTTPException\OKException();
+	}
+
+	protected function rawContent(array $request = [])
+	{
+		$nickname   = $this->parameters['nickname'] ?? '';
+		$contact_id = $this->parameters['cid']      ?? 0;
+
+		$hub_mode      = trim($request['hub_mode']         ?? '');
+		$hub_topic     = trim($request['hub_topic']        ?? '');
+		$hub_challenge = trim($request['hub_challenge']    ?? '');
+		$hub_verify    = trim($request['hub_verify_token'] ?? '');
+
+		$this->logger->notice('Subscription from ' . $this->server['REMOTE_ADDR'] . ' Mode: ' . $hub_mode . ' Nick: ' . $nickname);
```

Comment:
Include `src\App\Request` in this class and than:

```suggestion
		$this->logger->notice('Subscription start.', ['from' => $this->request->getRemoteAddress(), 'mode' => $hub_mode, 'nickname' => $nickname]);
```

Otherwise, using it with reverse proxies wouldn't show the right IP